In [1]:
import pandas as pd 
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv(r'D:\Downloadd\Harshada\DATA\Groceries.csv',header=None,names=['Products'])
print('shape of data',df.shape)
df.head()

shape of data (20, 1)


Products
0             MILK,BREAD,BISCUIT
1  BREAD,MILK,BISCUIT,CORNFLAKES
2            BREAD,TEA,BOURNVITA
3           JAM,MAGGI,BREAD,MILK
4              MAGGI,TEA,BISCUIT

In [3]:
data=list(df['Products'].apply(lambda x:x.split(',')))
data

[['MILK', 'BREAD', 'BISCUIT'],
 ['BREAD', 'MILK', 'BISCUIT', 'CORNFLAKES'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['JAM', 'MAGGI', 'BREAD', 'MILK'],
 ['MAGGI', 'TEA', 'BISCUIT'],
 ['BREAD', 'TEA', 'BOURNVITA'],
 ['MAGGI', 'TEA', 'CORNFLAKES'],
 ['MAGGI', 'BREAD', 'TEA', 'BISCUIT'],
 ['JAM', 'MAGGI', 'BREAD', 'TEA'],
 ['BREAD', 'MILK'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'COCK', 'BISCUIT', 'CORNFLAKES'],
 ['COFFEE', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'COCK'],
 ['BREAD', 'SUGER', 'BISCUIT'],
 ['COFFEE', 'SUGER', 'CORNFLAKES'],
 ['BREAD', 'SUGER', 'BOURNVITA'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['BREAD', 'COFFEE', 'SUGER'],
 ['TEA', 'MILK', 'COFFEE', 'CORNFLAKES']]

In [4]:
from mlxtend.preprocessing import TransactionEncoder
te =TransactionEncoder()
te_data=te.fit_transform(data)

df=pd.DataFrame(te_data,columns=te.columns_).astype(int)
df

BISCUIT  BOURNVITA  BREAD  COCK  COFFEE  CORNFLAKES  JAM  MAGGI  MILK  \
0         1          0      1     0       0           0    0      0     1   
1         1          0      1     0       0           1    0      0     1   
2         0          1      1     0       0           0    0      0     0   
3         0          0      1     0       0           0    1      1     1   
4         1          0      0     0       0           0    0      1     0   
5         0          1      1     0       0           0    0      0     0   
6         0          0      0     0       0           1    0      1     0   
7         1          0      1     0       0           0    0      1     0   
8         0          0      1     0       0           0    1      1     0   
9         0          0      1     0       0           0    0      0     1   
10        1          0      0     1       1           1    0      0     0   
11        1          0      0     1       1           1    0      0     0   
12        0          1      0     0       1           0    0      0     0   
13        0          0      1     1       1           0    0      0     0   
14        1          0      1     0       0           0    0      0     0   
15        0          0      0     0       1           1    0      0     0   
16        0          1      1     0       0           0    0      0     0   
17        0          0      1     0       1           0    0      0     0   
18        0          0      1     0       1           0    0      0     0   
19        0          0      0     0       1           1    0      0     1   

    SUGER  TEA  
0       0    0  
1       0    0  
2       0    1  
3       0    0  
4       0    1  
5       0    1  
6       0    1  
7       0    1  
8       0    1  
9       0    0  
10      0    0  
11      0    0  
12      1    0  
13      0    0  
14      1    0  
15      1    0  
16      1    0  
17      1    0  
18      1    0  
19      0    1

* Applying Apriori and Resulting

The next step is to create the Apriori Model. We can change all the parameters in the Apriori Model in the mlxtend package.
I will try to use minimum support parameters for this modeling.
For this, I set a min_support value with a threshold value of 20% and printed them on the screen as well.

In [5]:
from mlxtend.frequent_patterns import apriori, association_rules
#set a threshold value for the support value and calculate the support value
df=apriori(df,min_support=0.2,use_colnames=True,verbose=1)
df

Processing 42 combinations | Sampling itemset size 3


C:\Users\Admin\AppData\Roaming\Python\Python39\site-packages\mlxtend\frequent_patterns\fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support              itemsets
0      0.35             (BISCUIT)
1      0.20           (BOURNVITA)
2      0.65               (BREAD)
3      0.40              (COFFEE)
4      0.30          (CORNFLAKES)
5      0.25               (MAGGI)
6      0.25                (MILK)
7      0.30               (SUGER)
8      0.35                 (TEA)
9      0.20      (BISCUIT, BREAD)
10     0.20         (BREAD, MILK)
11     0.20        (BREAD, SUGER)
12     0.20          (TEA, BREAD)
13     0.20  (CORNFLAKES, COFFEE)
14     0.20       (COFFEE, SUGER)
15     0.20          (TEA, MAGGI)

* I chose the 60% minimum confidence value. In other words, when product X is purchased, we can say that the purchase of product Y is 60% or more.

In [6]:
#Let's view our interpretation values using the Associan rule function.
df_ar = association_rules(df, metric = "confidence", min_threshold = 0.6)
df_ar

antecedents consequents  antecedent support  consequent support  support  \
0        (MILK)     (BREAD)                0.25                0.65      0.2   
1       (SUGER)     (BREAD)                0.30                0.65      0.2   
2  (CORNFLAKES)    (COFFEE)                0.30                0.40      0.2   
3       (SUGER)    (COFFEE)                0.30                0.40      0.2   
4       (MAGGI)       (TEA)                0.25                0.35      0.2   

   confidence      lift  leverage  conviction  zhangs_metric  
0    0.800000  1.230769    0.0375        1.75       0.250000  
1    0.666667  1.025641    0.0050        1.05       0.035714  
2    0.666667  1.666667    0.0800        1.80       0.571429  
3    0.666667  1.666667    0.0800        1.80       0.571429  
4    0.800000  2.285714    0.1125        3.25       0.750000

For example, if we examine our 1st index value;

* The probability of seeing sugar sales is seen as 30%.
* Bread intake is seen as 65%.
* We can say that the support of both of them is measured as 20%.
* 67% of those who buys sugar, buys bread as well.
* Users who buy sugar will likely consume 3% more bread than users who don't buy sugar.
* Their correlation with each other is seen as 1.05.
* As a result, if item X and Y are bought together more frequently, then several steps can be taken to increase the profit. For instance:
